pkg/proxy/util/service.go

```diff
@@ -0,0 +1,145 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"reflect"
+	"sync"
+
+	"github.com/golang/glog"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/proxy"
+)
+
+// ProxyServiceMap maps a service to its service info.
+type ProxyServiceMap map[proxy.ServicePortName]proxy.ServiceInfo
+
+// endpointsChange stores both the previous and current ProxyServiceMap.
+type serviceChange struct {
+	// previous ProxyServiceMap.
+	previous ProxyServiceMap
+	// current ProxyServiceMap.
+	current ProxyServiceMap
+}
+
+// ServiceChangeMap maps a service to its serviceChange.
+type ServiceChangeMap struct {
+	// lock protects items.
+	lock sync.Mutex
+	// items is a service to its serviceChange map.
+	items map[types.NamespacedName]*serviceChange
+}
+
+func NewServiceChangeMap() ServiceChangeMap {
+	return ServiceChangeMap{
+		items: make(map[types.NamespacedName]*serviceChange),
+	}
+}
+
+// Merge adds other ProxyEndpointsMap's elements to current ProxyEndpointsMap, if exists, update it.
+// It returns all other ProxyEndpointsMap's service ports.
+func (sm *ProxyServiceMap) Merge(other ProxyServiceMap) sets.String {
+	existingPorts := sets.NewString()
+	for svcPortName, info := range other {
+		existingPorts.Insert(svcPortName.Port)
+		_, exists := (*sm)[svcPortName]
+		if !exists {
+			glog.V(1).Infof("Adding new service port %q at %s:%d/%s", svcPortName, info.ClusterIP(), info.Port(), info.Protocol())
+		} else {
+			glog.V(1).Infof("Updating existing service port %q at %s:%d/%s", svcPortName, info.ClusterIP(), info.Port(), info.Protocol())
+		}
+		(*sm)[svcPortName] = info
+	}
+	return existingPorts
+}
+
+// Unmerge deletes current ProxyServiceMap's elements which are in other ProxyEndpointsMap.
```

Comment:
Yeah. It's fixed now and comment message is 

'// Unmerge deletes all other ServiceMap's elements from current ServiceMap.  The returned value of Merge() is the `existingPorts` argument of Unmerge().
// If a service identifier exists in existingPorts, it means it's an `Update` action and will skip delete, otherwise it's a `Delete` action and will delete the service from ServiceMap.
// We pass in the `UDPStaleClusterIP`argument for storing the stale udp service cluster IP and clear stale udp connection later.'

BTW, I also think
`func (sm *ServiceMap) Unmerge(other ServiceMap, existingPorts, UDPStaleClusterIP sets.String)` is a weird function signature and we can refactor the `UDPStaleClusterIP` to be the returned value later. wdyt?